File: battle/ELineUp.java
package common.battle;

import common.CommonStatic;
import common.util.BattleObj;
import common.util.stage.Limit;
import common.util.stage.StageLimit;
import common.util.unit.Form;

import java.util.Arrays;

public class ELineUp extends BattleObj {

	public final int[][] price, cool, maxC, tick, cdDownOrb, priceDownOrb;
	private final StageBasis b;

	protected ELineUp(LineUp lu, StageBasis sb) {
		b = sb;
		price = new int[2][5];
		cool = new int[2][5];
		maxC = new int[2][5];
		tick = new int[2][5];
		cdDownOrb = new int[2][5];
		priceDownOrb = new int[2][5];
		Limit lim = sb.est.lim;
		for (int i = 0; i < 2; i++)
			for (int j = 0; j < 5; j++) {
				Form form = lu.fs[i][j];
				if (form == null) {
					price[i][j] = -1;
					continue;
				}
				if (lim != null && ((lim.line == 1 && i == 1) || lim.unusable(lu.efs[i][j].du, sb.st.getCont().price)))
					price[i][j] = -1;
				else
					price[i][j] = 100 * (sb.globalCost() > -1 ? sb.globalCost() : (int) (lu.efs[i][j].getPrice(sb.st.getCont().price)));
				if (!StageLimit.isComboBanned(lim, C_DISCOUNT))
					price[i][j] -= price[i][j] * b.b.getInc(C_DISCOUNT, form.du.getPack().unit) / 100;
				maxC[i][j] = sb.globalCdLimit() > 0
						? sb.b.t().getFinResGlobal(sb.globalCdLimit(), StageLimit.isComboBanned(sb.est.lim, C_RESP) ? 0 : sb.b.getInc(C_RESP, lu.efs[i][j].du.getPack().unit))
						: sb.b.t().getFinRes(lu.efs[i][j].du.getRespawn(), StageLimit.isComboBanned(sb.est.lim, C_RESP) ? 0 : sb.b.getInc(C_RESP, lu.efs[i][j].du.getPack().unit));
				if (lim != null && lim.stageLimit != null) {
					if (price[i][j] != -1)
						price[i][j] = price[i][j] * lim.stageLimit.costMultiplier[form.unit.rarity] / 100;
					maxC[i][j] = maxC[i][j] * lim.stageLimit.cooldownMultiplier[form.unit.rarity] / 100;
				}
				int[][] orbs = lu.efs[i][j].getLevel().getOrbs();
				boolean hasEveryOther = false;
				if (orbs != null) {
					for (int[] orb : orbs) {
						if (orb.length != ORB_INTS)
							continue;
						int orbId = orb[0];
						hasEveryOther |= Arrays.stream(ORB_EVERY_OTHER).anyMatch(v -> v == orbId);
						if (orbId == ORB_COOLDOWN)
							cdDownOrb[i][j] = ORB_COOLDOWN_MULT[orb[2]];
						else if (orbId == ORB_COST_DOWN)
							priceDownOrb[i][j] = ORB_COST_DOWN_MULT[orb[2]];
					}
				}
				if (!hasEveryOther)
					tick[i][j] = -1;
			}
	}

	/**
	 * reset cooldown of a unit
	 */
	protected void get(int i, int j) {
		cool[i][j] = maxC[i][j];
		if (cdDownOrb[i][j] > 0 && tick[i][j] == 0)
			cool[i][j] -= cool[i][j] * cdDownOrb[i][j] / 100;
		b.cdDelayVisual[i][j] = StageBasis.DELAY_BASE.clone();
	}

	protected void delay(int i, int j, int[] delay) {
		if (cool[i][j] == 0)
			return;

		int inc = b.getDelayStrength(cool[i][j], maxC[i][j], delay);
		if (inc > 0) {
			b.cdDelayVisual[i][j][0] = Math.max(b.cdDelayVisual[i][j][0], cool[i][j]);
		} else {
			b.cdDelayVisual[i][j][2] += inc;
		}
		cool[i][j] += inc;
		if (cool[i][j] > maxC[i][j])
			cool[i][j] = maxC[i][j];
		if (inc < 0) {
			if (cool[i][j] <= 0) {
				cool[i][j] = 0;
				CommonStatic.setSE(SE_SPEND_REF);
				b.frameOffCd[i][j] = b.time;
			} else {
				b.cdDelayVisual[i][j][3] = 10;
			}
			CommonStatic.setSE(SE_DELAY_COOLDOWN);
		} else {
			b.cdDelayVisual[i][j][1] = 10;
			CommonStatic.setSE(SE_DELAY_COOLDOWN);
		}
	}

	/**
	 * count down the cooldown
	 */
	protected void update() {
		for (int i = 0; i < 2; i++)
			for (int j = 0; j < 5; j++) {
				if (cool[i][j] > 0) {
					cool[i][j]--;

//					if (cool[i][j] == 30)
//						delay(i, j, -30, 0);

					if (cool[i][j] == 0) {
						CommonStatic.setSE(SE_SPEND_REF);
						b.frameOffCd[i][j] = b.time;
					}
				}
				if (b.cdDelayVisual[i][j][1] > 0 && --b.cdDelayVisual[i][j][1] == 0) {
					b.cdDelayVisual[i][j][0] = 0;
				}
				if (b.cdDelayVisual[i][j][3] > 0 && --b.cdDelayVisual[i][j][3] == 0) {
					b.cdDelayVisual[i][j][2] = 0;
				}
			}
	}

}

File: battle/ELineUp.kt
package common.battle

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.BattleObj
import common.util.stage.EStage
import common.util.stage.Limit
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter

class ELineUp(lu: LineUp, sb: StageBasis) : BattleObj() {
    private val b: StageBasis
    val price: Array<IntArray>
    val cool: Array<IntArray>
    val maxC: Array<IntArray>

    /** reset cool down time of an unit  */
    operator fun get(i: Int, j: Int) {
        cool[i][j] = maxC[i][j]
    }

    /** count down the cool down time  */
    fun update() {
        for (i in 0..1) for (j in 0..4) if (cool[i][j] > 0) cool[i][j]--
    }

    init {
        b = sb
        price = Array(2) { IntArray(5) }
        cool = Array(2) { IntArray(5) }
        maxC = Array(2) { IntArray(5) }
        val lim: Limit = b.est.lim
        for (i in 0..1) for (j in 0..4) {
            if (lu.fs.get(i).get(j) == null) {
                price[i][j] = -1
                continue
            }
            price[i][j] = lu.efs.get(i).get(j).getPrice(b.st.map.price)
            maxC[i][j] = sb.b.t().getFinRes(lu.fs.get(i).get(j).du.getRespawn())
            if (lim == null) continue
            val overl = lim.rare shr lu.fs.get(i).get(j).unit.rarity
            if (lim.rare != 0 && overl and 1 == 0) price[i][j] = -1
            if (lim.line == 1 && i == 1) price[i][j] = -1
            if (lim.min > 0 && price[i][j] < lim.min) price[i][j] = -1
            if (lim.max > 0 && price[i][j] > lim.max) price[i][j] = -1
            if (lim.group != null && !lim.group.allow(lu.fs.get(i).get(j).unit)) price[i][j] = -1
        }
    }
}
